From 5fd491b49a69f933b529aa5d8f7e0a37c06799bb Mon Sep 17 00:00:00 2001
From: Vladimir Sokolovsky <vlad@mellanox.com>
Date: Mon, 25 Jun 2012 22:28:24 +0300
Subject: [PATCH 03/21] BACKPORT: core/netlink for kernels <= 3.4

Signed-off-by: Vladimir Sokolovsky <vlad@mellanox.com>
---
 drivers/infiniband/core/netlink.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/infiniband/core/netlink.c b/drivers/infiniband/core/netlink.c
index e497dfb..2ac867a 100644
--- a/drivers/infiniband/core/netlink.c
+++ b/drivers/infiniband/core/netlink.c
@@ -149,12 +149,18 @@ static int ibnl_rcv_msg(struct sk_buff *skb, struct nlmsghdr *nlh)
 			    !client->cb_table[RDMA_NL_GET_OP(op)].dump)
 				return -EINVAL;
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,4,0))
 			{
 				struct netlink_dump_control c = {
 					.dump = client->cb_table[op].dump,
 				};
 				return netlink_dump_start(nls, skb, nlh, &c);
 			}
+#else
+			return netlink_dump_start(nls, skb, nlh,
+						  client->cb_table[op].dump,
+						  NULL, 0);
+#endif /* (LINUX_VERSION_CODE < KERNEL_VERSION(3,4,0)) */
 		}
 	}
 
-- 
1.7.0.4

