commit cba9df92717abd6d00e758c79fe11af97f15eed9
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Nov 19 14:44:44 2014 -0600

    cxgb3/4: add "-ofed" suffix to driver version

diff --git a/drivers/net/ethernet/chelsio/cxgb3/version.h b/drivers/net/ethernet/chelsio/cxgb3/version.h
index 165bfb9..1cfffe2 100644
--- a/drivers/net/ethernet/chelsio/cxgb3/version.h
+++ b/drivers/net/ethernet/chelsio/cxgb3/version.h
@@ -35,7 +35,7 @@
 #define DRV_DESC "Chelsio T3 Network Driver"
 #define DRV_NAME "cxgb3"
 /* Driver version */
-#define DRV_VERSION "1.1.5-ko"
+#define DRV_VERSION "1.1.5-ofed"
 
 /* Firmware version */
 #define FW_VERSION_MAJOR 7
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
index 578f7cd..ea34737 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
@@ -77,7 +77,7 @@
 #ifdef DRV_VERSION
 #undef DRV_VERSION
 #endif
-#define DRV_VERSION "2.0.0-ko"
+#define DRV_VERSION "2.0.0-ofed"
 #define DRV_DESC "Chelsio T4/T5 Network Driver"
 
 /*
