--- compat-rdma/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c	2014-08-28 15:32:56.316456204 -0500
+++ compat-rdma.new/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c	2014-08-28 15:28:29.061456300 -0500
@@ -4693,6 +4693,18 @@ static int cxgb_set_mac_addr(struct net_
 	return 0;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,1,0)
+static void vlan_rx_register(struct net_device *dev, struct vlan_group *grp)
+{
+        struct port_info *pi = netdev_priv(dev);
+        struct adapter *adapter = pi->adapter;
+
+        pi->vlan_grp = grp;
+        t4_set_rxmode(adapter, adapter->mbox, pi->viid, -1, -1, -1, -1,
+                      grp != NULL, 0);
+}
+#endif
+
 #ifdef CONFIG_NET_POLL_CONTROLLER
 static void cxgb_netpoll(struct net_device *dev)
 {
@@ -4730,6 +4742,9 @@ static const struct net_device_ops cxgb4
 #ifdef CONFIG_NET_POLL_CONTROLLER
 	.ndo_poll_controller  = cxgb_netpoll,
 #endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,1,0)
+        .ndo_vlan_rx_register   = vlan_rx_register,
+#endif
 };
 
 void t4_fatal_err(struct adapter *adap)
