From 4f27d323bd47563f40a663672a331c5b2c95138e Mon Sep 17 00:00:00 2001
From: Phil Cayton <phil.cayton@intel.com>
Date: Tue, 4 Feb 2014 12:25:45 -0800
Subject: [PATCH 09/12] update drivers/infiniband's Kconfig and Makefile to
 allow compilation of CCL-Direct (ibp)

Signed-off-by: Phil Cayton <phil.cayton@intel.com>
---
diff -urN a8/drivers/infiniband/hw/Makefile a9/drivers/infiniband/hw/Makefile
--- a8/drivers/infiniband/hw/Makefile	2015-01-05 15:04:13.993463721 -0800
+++ a9/drivers/infiniband/hw/Makefile	2015-01-05 15:09:10.056451249 -0800
@@ -10,3 +10,4 @@
 obj-$(CONFIG_INFINIBAND_NES)		+= nes/
 obj-$(CONFIG_INFINIBAND_OCRDMA)		+= ocrdma/
 obj-$(CONFIG_INFINIBAND_USNIC)		+= usnic/
+obj-$(CONFIG_INFINIBAND_SCIF)		+= scif/
diff -urN a8/drivers/infiniband/Kconfig a9/drivers/infiniband/Kconfig
--- a8/drivers/infiniband/Kconfig	2015-01-05 15:04:14.001463720 -0800
+++ a9/drivers/infiniband/Kconfig	2015-01-05 15:07:03.176456594 -0800
@@ -55,6 +55,9 @@
 source "drivers/infiniband/hw/nes/Kconfig"
 source "drivers/infiniband/hw/ocrdma/Kconfig"
 source "drivers/infiniband/hw/usnic/Kconfig"
+source "drivers/infiniband/hw/scif/Kconfig"
+
+source "drivers/infiniband/ibp/Kconfig"
 
 source "drivers/infiniband/ulp/ipoib/Kconfig"
 
diff -urN a8/drivers/infiniband/Makefile a9/drivers/infiniband/Makefile
--- a8/drivers/infiniband/Makefile	2015-01-05 15:04:14.001463720 -0800
+++ a9/drivers/infiniband/Makefile	2015-01-05 15:08:25.112453143 -0800
@@ -1,3 +1,4 @@
 obj-$(CONFIG_INFINIBAND)		+= core/
 obj-$(CONFIG_INFINIBAND)		+= hw/
 obj-$(CONFIG_INFINIBAND)		+= ulp/
+obj-$(CONFIG_IBP_SERVER)		+= ibp/
