Index: compat-rdma/drivers/infiniband/ulp/iser/iscsi_iser.c
===================================================================
--- compat-rdma.orig/drivers/infiniband/ulp/iser/iscsi_iser.c
+++ compat-rdma/drivers/infiniband/ulp/iser/iscsi_iser.c
@@ -719,7 +719,9 @@ static struct iscsi_transport iscsi_iser
 	.create_conn            = iscsi_iser_conn_create,
 	.bind_conn              = iscsi_iser_conn_bind,
 	.destroy_conn           = iscsi_iser_conn_destroy,
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(3,2,0))
 	.attr_is_visible	= iser_attr_is_visible,
+#endif
 	.set_param              = iscsi_iser_set_param,
 	.get_conn_param		= iscsi_conn_get_param,
 	.get_ep_param		= iscsi_iser_get_ep_param,
