From 13a3b85af5ae8e6dce4fb988f7b43839a867be3a Mon Sep 17 00:00:00 2001
From: Vladimir Sokolovsky <vlad@mellanox.com>
Date: Thu, 25 Jul 2013 19:50:07 +0300
Subject: [PATCH] cxgb4/cm: Fix compilation on SLES11.3

Signed-off-by: Vladimir Sokolovsky <vlad@mellanox.com>
---
 drivers/infiniband/hw/cxgb4/cm.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index 2271a7f..7a9e75f 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -1601,6 +1601,8 @@ static int c4iw_reconnect(struct c4iw_ep *ep)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 1, 0)
 	neigh = dst_neigh_lookup(ep->dst,
 			&ep->com.cm_id->remote_addr.sin_addr.s_addr);
+#elif defined(CONFIG_COMPAT_SLES_11_3)
+	neigh = ep->dst->_neighbour;
 #else
 	neigh = ep->dst->neighbour;
 #endif
@@ -3107,6 +3109,8 @@ static int rx_pkt(struct c4iw_dev *dev, struct sk_buff *skb)
 	neigh = dst_neigh_lookup_skb(dst, skb);
 #elif LINUX_VERSION_CODE >= KERNEL_VERSION(3, 1, 0)
 	neigh = dst->_neighbour;
+#elif defined(CONFIG_COMPAT_SLES_11_3)
+	neigh = dst->_neighbour;
 #else
 	neigh = dst->neighbour;
 #endif
-- 
1.7.9.5

